lib/retry.go

```diff
@@ -0,0 +1,83 @@
+package lib
+
+import (
+	"time"
+)
+
+const (
+	defaultMinFailures = 0
+	defaultMaxWait     = 2 * time.Minute
+)
+
+// RetryWaiter will record failed and successful operations and provide
+// a channel to wait on before a failed operation can be retried.
+type RetryWaiter struct {
+	minFailures uint
+	maxWait     time.Duration
+	failures    uint
+}
+
+// Creates a new RetryWaiter
+func NewRetryWaiter(minFailures int, maxWait time.Duration) *RetryWaiter {
+	if minFailures < 0 {
+		minFailures = defaultMinFailures
+	}
+
+	if maxWait <= 0 {
+		maxWait = defaultMaxWait
+	}
+
+	return &RetryWaiter{
+		minFailures: uint(minFailures),
+		maxWait:     maxWait,
+		failures:    0,
+	}
+}
+
+// Marks that an operation is successful which resets the failure count.
+// The chan that is returned will be immediately selectable
+func (rw *RetryWaiter) Success() <-chan time.Time {
+	rw.Reset()
+	c := make(chan time.Time, 1)
+	c <- time.Now()
+	return c
+}
+
+// Marks that an operation failed. The chan returned will be selectable
+// once the calculated retry wait amount of time has elapsed
+func (rw *RetryWaiter) Failed() <-chan time.Time {
```

Comment:
Is there a good reason to make this a `chan time.Time`?

It seems like we don't expect the caller to ever care about the actual returned time value so should it be a `chan struct{}`? It's marginally cleaner API, and it allows minor improvements like `Success` can just return a closed chan rather than needing to buffer we could even keep a single closed chan around from `New` and return the same one to ever success call so there is no allocation on the happy path but that's a microoptimisation.

It would only add a few extra lines to turn `time.After` into `time.AfterFunc(waitTime, func(){ close(ch) })`.

Not a big deal but seems slightly nicer to me...

Reply:
It certainly could be a chan struct. The main reason I did it was to hand back the time.After chan but your way works too. I can change it.